cdap-ui/app/features/flows/controllers/tabs/runs/flowlets/input-ctrl.js

```diff
@@ -25,152 +28,159 @@ angular.module(PKG.name + '.feature.flows')
+        '&tag=app:' + $state.params.appId +
+        '&tag=flow' + $state.params.programId +
+        '&tag=flowlet:' + flowletid +
+        '&tag=run:' + $scope.runs.selected.runid +
+        '&start=now-1s&end=now';
+      // TODO: should this value be averaged over more than just the past 1 second?
+      // POLLING ARRIVAL RATE
+      dataSrc
+        .poll({
+          _cdapPath: arrivalPath,
+          method: 'POST'
+        }, function (res) {
+          if (res.series[0]) {
+            input.total = res.series[0].data[0].value;
+          }
+        });
+    }
+
+    function formatInitialTimeseries(aggregate, initial, input) {
+      var response = initial.series[0].data;
+      var v = [];
+
+      response[response.length - 1].value = aggregate - response[response.length - 1].value;
+      for (var i = response.length - 2; i >= 0; i--) {
+        response[i].value = response[i+1].value - response[i].value;
+        v.unshift({
+          time: response[i].time,
+          y: response[i].value
+        });
+      }
 
-          });
+      input.stream = v.slice(-1);
 
+      input.history = [
+        {
+          label: 'output',
+          values: v
+        }
+      ];
+    }
+
+    function pollAggregateQueue(path, input) {
+      dataSrc.poll({
+        _cdapPath: path + '&start=now-60s&end=now&aggregate=true',
+        method: 'POST',
+        interval: 1000
+      }, function(streamData) {
+        var stream = {};
+
+        if (streamData.series[0]) {
+          stream = {
+            time: Math.floor((new Date()).getTime()/1000),
+            y: streamData.series[0].data[0].value
+          };
+        } else {
+          stream = {
+            time: Math.floor((new Date()).getTime()/1000),
+            y: 0
+          };
         }
 
+        var array = input.history[0].values;
+        array.shift();
+        array.push(stream);
+
+        input.history = [
+          {
+            label: 'output',
+            values: array
+          }
+        ];
+
+        input.stream = array.slice(-1);
+        input.max = Math.max.apply(Math, array.map(function(o){return o.y;}));
+
       });
+    }
+
+
+    function formatInput() {
+      angular.forEach($scope.inputs, function (input) {
+        var flowletTags = {
+          namespace: $state.params.namespace,
+          app: $state.params.appId,
+          flow: $state.params.programId,
+          consumer: flowletid,
+          producer: input.name
+        };
+
+        var path = '/metrics/query?' + MyMetricsQueryHelper.tagsToParams(flowletTags)
+                          + '&metric=system.queue.pending';
+
+        var aggregate = 0;
+        // Get Aggregate
+        dataSrc.request({
+          _cdapPath: path + '&start=now-60s&end=now&aggregate=true',
+          method: 'POST'
+        }).then(function(res) {
+
+          // Get initial aggregate
+          aggregate = res.series[0] ? res.series[0].data[0].value : 0;
+
+        }).then(function() {
+
```

Comment:
Since we are returning any promise in the previous ```then``` handler would this then function even be executed? The idea of chaining promises is to return back a promise from each then clause so that the next subsequent then function will be executed right?